File: 49. Group Anagrams/49. Group Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const anagrams = new Map();

  for (const s of strs) {
    const alphabet = new Array(26).fill(0);
    for (const c of s) {
      alphabet[c.charCodeAt(0) - "a".charCodeAt(0)]++;
    }
    const key = alphabet.join(",");
    if (!anagrams.get(key)) {
      anagrams.set(key, []);
    }
    anagrams.get(key).push(s);
  }

  return Array.from(anagrams.values());
};

File: 49. Group Anagrams/49. Group Anagrams.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)

        for s in strs:
            alphabet = [0] * 26
            for c in s:
                alphabet[ord(c) - ord("a")] += 1
            anagrams[tuple(alphabet)].append(s)

        return list(anagrams.values())
